IMPORT NECCESSARY PACKAGES

In [4]:
import sys
import pandas as pd
import numpy as np
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error



CLEAN THE DATASET AND PREPARE IT FOR BUILDING A MODEL

In [17]:
# Load the dataset
file_path = 'data/venues_bikes_df.csv'
df = pd.read_csv(file_path)

# Drop unnecessary columns: 'latitude', 'longitude', 'business_name', 'phone', 'address'
df_cleaned = df.drop(columns=['latitude_y', 'longitude_y','latitude_x', 'longitude_x', 'business_name', 'phone', 'address', 'source','station_name'])

# Remove rows with blank 'price'
df_cleaned = df_cleaned[df_cleaned['price'].notna()]

# Map the price symbols to numerical values: $ -> 1, $$ -> 2, $$$ -> 3, $$$$ -> 4
price_mapping = {'$': 1, '$$': 2, '$$$': 3, '$$$$': 4}
df_cleaned['price'] = df_cleaned['price'].map(price_mapping)


# One-hot encode the 'categories' column (split if there are multiple categories per row)
df_cleaned = df_cleaned.join(df_cleaned['categories'].str.get_dummies(sep=', '))

# Drop the original 'categories' column after encoding
df_cleaned = df_cleaned.drop(columns=['categories'])


# Display the cleaned DataFrame
df_cleaned.head()


free_bikes  ebikes  empty_slots  review_count  rating  price    distance  \
0           1       1           33         254.0     4.3      2  178.845344   
1           1       1           33         686.0     4.2      2  169.517456   
2           1       1           33         232.0     4.2      2  191.044234   
3           1       1           33         123.0     4.3      3  152.157897   
4           1       1           33          61.0     4.2      3  168.615262   

   Afghan  African  American  ...  Ukrainian  Vegan  Vegetarian  \
0       0        0         0  ...          0      0           0   
1       0        0         0  ...          0      0           0   
2       0        0         0  ...          0      0           0   
3       0        0         0  ...          0      0           0   
4       0        0         0  ...          0      0           0   

   Venues & Event Spaces  Vietnamese  Waffles  Whiskey Bars  Wine & Spirits  \
0                      0           0        0             0               0   
1                      0           0        0             0               0   
2                      0           0        0             0               0   
3                      0           0        0             0               0   
4                      0           0        0             0               0   

   Wine Bars  Wine Tours  
0          0           0  
1          0           0  
2          0           0  
3          0           0  
4          0           0  

[5 rows x 146 columns]

In [18]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Step 1: Prepare the Data
X = df_cleaned.drop(columns=['rating'])  # Features (all columns except 'rating')
y = df_cleaned['rating']  # Target variable

# Step 2: Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Train the Model (Linear Regression)
model = LinearRegression()
model.fit(X_train, y_train)

# Step 4: Make Predictions
y_pred = model.predict(X_test)

# Step 5: Evaluate the Model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")


Mean Squared Error: 0.09600072709128776
R-squared: 0.2815735310062012


Mean Squared Error (MSE): 0.096: This represents the average squared difference between the predicted ratings and the actual ratings.

R-squared (R²): 0.282: This means that about 28.2% of the variability in the rating can be explained by your features. While this suggests that the model is capturing some relationship between the features and the target variable, the relatively low R² also indicates that there is significant variability in the rating that is not being captured by the model.

In [ ]:
#Stretch
#How can you turn the regression model into a classification model?